src/Lean/Meta/Tactic/FunInd.lean

```diff
@@ -0,0 +1,855 @@
+/-
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joachim Breitner
+-/
+
+prelude
+import Lean.Meta.Basic
+import Lean.Meta.Match.MatcherApp.Transform
+import Lean.Meta.Check
+import Lean.Meta.Tactic.Cleanup
+import Lean.Meta.Injective -- for elimOptParam
+import Lean.Elab.PreDefinition.WF.Eqns
+import Lean.Elab.PreDefinition.WF.PackMutual
+import Lean.Elab.Command
+
+/-!
+This module contains code to derive, from the definition of a (possibly mutual) recursive function,
+a **functional induction principle** tailored to proofs about that function. For example from:
+
+```
+def ackermann : Nat → Nat → Nat
+  | 0, m => m + 1
+  | n+1, 0 => ackermann n 1
+  | n+1, m+1 => ackermann n (ackermann (n + 1) m)
+derive_induction ackermann
+```
+we get
+```
+ackermann.induct (motive : Nat → Nat → Prop) (case1 : ∀ (m : Nat), motive 0 m)
+  (case2 : ∀ (n : Nat), motive n 1 → motive (Nat.succ n) 0)
+  (case3 : ∀ (n m : Nat), motive (n + 1) m → motive n (ackermann (n + 1) m) → motive (Nat.succ n) (Nat.succ m))
+  (x x : Nat) : motive x x
+```
+
+## Spec
+
+The functional induction principle
+
+* Take the same fixed parameters as the function.
+* The motive takes the same non-fixed parameter as the original function.
+* For each branch of the original function, there is a case in the induction principle.
+  Here branch roughly corresponds to tail-call positions: Branches of top-level
+  if-then-else and matches.
+* The local context of the branch (e.g. the condition of an if-then-else) is provided as assumptions
+  in the case.
+* In addition, for every recursive call in that branch, an induction hypothesis asserting the motive
+  for the arguments of the recursive call is provided.
+* If the recursive call is under binder and it, or its proof of termination, on the the bound,
+  values, these become assumptions on the inductive hypothesis.
+* Mutual recursion is supported and results in multiple motives.
+
+
+## Implementation overview
+
+For a non-mutual, unary function `foo` (or else for the `_unary` function), we
+
+1. expect its definition, possibly after some `whnf`’ing, to be of the form
+   ```
+   def foo := fun x₁ … xₙ (y : a) => fix (fun y' oldIH => body) y
+   ```
+   where `xᵢ…` are the fixed prefix and `y` is parameter of the function.
+
+2. From this structure we derive the type of the `motive`, and start assembling the induction
+   principle:
+   ```
+   def foo.induct := fun x₁ … xₙ (motive : (y : a) → Prop) =>
+    fix (fun y' newIH => T[body])
+   ```
+
+3. The first phase, transformation `T1[body]` (implemented in) `buildInductionBody`,
+   mirrors the branching structure of `foo`, i.e. replicates `ite` and matcher applications,
```

Comment:
What about `dite`?

Reply:
```suggestion
   mirrors the branching structure of `foo`, i.e. replicates `dite` and matcher applications,
```
(In fact, there is no `ite` here because `Fix.lean` replaces `ite` with `dite` in the body of recursive functions.)